### ISSUE TITLE
[V3] Remove the publisher interface

### ISSUE THREAD

**Issue by mefellows** [pact-foundation/pact-js#765]
One of the problems with bundling the Ruby binaries into the V3 package is that we still have the long path issue for our Windows friends (see this redacted reach out from a Pactflow customer):

![Screen Shot 2021-11-02 at 10 45 59 am](https://user-images.githubusercontent.com/53900/139756844-6edbc566-3f39-419b-a9a4-43f405a4f72f.png)

If we are to encourage the use of the CLI over the NodeJS interface, we should remove the `Publisher` interface from Pact JS.

This is separate to bundling the Ruby CLI, which we probably ought to consider removing in the fullness of time also.

**Comment by TimothyJones** [comment]
Hmm. The cli will still have the windows long path issue.

**Comment by mefellows** [comment]
Yes, but the user can configure where it is installed, or can choose to use the Ruby gem or the docker container.

Bundled into the JS module is the primary reason why the long path issue is an issue.

**Comment by TimothyJones** [comment]
We’ll have the same problem if they’re using the binstubs. I think it would be better to detect the issue and suggest they fix it with one of those workarounds.

**Comment by TimothyJones** [comment]
Removing the publisher interface would just move the problems around for affected users, and introduce churn for the (majority) of users who aren't affected and still have the old `publish.js` pattern. 

I'm all for encouraging people to use the CLI in the future, but I don't think that removing the publisher interface is the best way to do that.
 
How about- 

To encourage users to use the CLI:

1) warn when the publisher interface is used from pact-js (and provide an option to silence this warning)
2) Stop exposing it from pact-js, but keep exposing it from pact-core

To address the windows path issue problem:

1) Allow the location of the binary to be configured in pact-core (I don't think we do that anymore)

Other options - how complex is the publisher? If it's just an API call, we could implement it natively.

**Comment by TimothyJones** [comment]
>  can choose to use the Ruby gem

The problem is with Ruby on windows. Using the Ruby gem will still have this issue.

**Comment by mefellows** [comment]
> Removing the publisher interface would just move the problems around for affected users, and introduce churn for the (majority) of users who aren't affected and still have the old `publish.js` pattern.

Yes that's a good point, but if we are to remove it at some point a major version bump is the time to do it. Should we consider deprecating it in the current 9.x.x now?

> I'm all for encouraging people to use the CLI in the future, but I don't think that removing the publisher interface is the best way to do that.
> 
> How about-
> 
> To encourage users to use the CLI:
> 
> 1. warn when the publisher interface is used (and provide an option to silence this warning)
> 2. Stop exposing it from pact-js, but keep exposing it from pact-core
> 
> To address the windows path issue problem:
> 
> 1. Allow the location of the binary to be configured in pact-core (I don't think we do that anymore)
> 
> Other options - how complex is the publisher? If it's just an API call, we could implement it natively.

Yep, I think so too. #2 is what this issue was specifically raised. I think #1 can be done in the current API and not wait for 10.x.x, where it will be removed entirely (FWIW all they need to do is import `pact-core` directly and their publisher script will be fixed, so whilst it's a breaking change, it's easy to retain if they want it without encouraging new users).

> We’ll have the same problem if they’re using the binstubs. I think it would be better to detect the issue and suggest they fix it with one of those workarounds.

Yes that's probably not a bad idea. I'm still unsure whether or not we should take the stance of including those binstubs long term as well.

> The problem is with Ruby on windows. Using the Ruby gem will still have this issue.

No that's not really true, at least no more than any other Ruby gem. Because of the way the Ruby standalone bundles an entire extra tree (for Ruby runtime) the directory hierarchy.

Also, if using the newer versions of Ruby it will have the Windows fix applied etc.  Basically, it's now just a Ruby problem.

**Comment by TimothyJones** [comment]
>  I'm still unsure whether or not we should take the stance of including those binstubs long term as well.

I think we should make pact-js as easy to use as possible, so being able to say "oh, the CLI is available in your npm scripts now" is a clear win. Otherwise users will have to go and download the CLI on their own, etc.

> Because of the way the Ruby standalone bundles an entire extra tree (for Ruby runtime) the directory hierarchy.

I suppose it might be true that not using the standalone could have a shorter path length, but that still feels like a workaround.

> Also, if using the newer versions of Ruby it will have the Windows fix applied etc.

Are you sure? I can't see it in any of the Ruby changelogs - and as far as I know, it's still not fixed in 3.0.2 (It looks like Ruby fixed it in their master 6 months ago, so I'm not sure why it wasn't released in 3.0.2).

**Comment by TimothyJones** [comment]
> This is separate to bundling the Ruby CLI, which we probably ought to consider removing in the fullness of time also.

Is there a planned replacement?

**Comment by mefellows** [comment]
> Are you sure? I can't see it in any of the Ruby changelogs - and as far as I know, it's still not fixed in 3.0.2 (It looks like Ruby fixed it in their master 6 months ago, so I'm not sure why it wasn't released in 3.0.2).

You might be right, I thought I saw that it was fixed but like said perhaps it hasn't actually landed in real software yet. The point is, it's a general Windows + Ruby problem that has existed for 20 years, and we made it worse by packing it into a nested node module directory and expanding an entire ruby library tree on top of it. It's way less likely to happen if they install it in other more standard ways. 

> I think we should make pact-js as easy to use as possible, so being able to say "oh, the CLI is available in your npm scripts now" is a clear win. Otherwise users will have to go and download the CLI on their own, etc.

That's fair. I'm not against providing options. It's not how we use the CLI (I almost always preferentially use docker for these sorts of tasks) but that doesn't mean it's wrong.

> Is there a planned replacement?

No, the Ruby CLI will be the primary CLI for the foreseeable future. There is a lot of capability in there that isn't worth re-writing in another language. I think Beth is simplifying some of the APIs as things evolve (Publishing is one of those which is currently several API calls that will be reduced to one), but I can't see the value in doing it given that the CLI should work on every platform right now.

**Comment by TimothyJones** [comment]
> and we made it worse by packing it into a nested node module directory 

This doesn't feel fair to me. Possibly adding `/node_modules/@pact-foundation/pact-node/standalone` might make it slightly more likely to trigger for users who are close to the threshold, but relative to users downloading it to their project directory (which is where you'd want to do it naturally), I don't think it's a big difference.

As you point out, this is a Ruby / travelling-ruby / pact-standalone issue - I think this means that support for windows won't be improved by removing the publisher interface - we'll still have the same problems.

If the plan is to use and extend the ruby standalone for the future, then I think we'll need to solve windows support in the standalone.

**Comment by mefellows** [comment]
I have removed the Publisher interface in the 10.x.x branch. The `publishPacts` operation still exists in https://github.com/pact-foundation/pact-js-core/#pact-broker-publishing, and the `Publisher` was simply a wrapper of that utility. 

I think you could make the argument that the `Publisher` class wasn't a big deal, but equally, you could make the case that we should also support `canDeploy` et. al. It's also not much of a leap to use the source library, and it removes the only remaining Ruby API dependency Pact JS surfaces (the rest are now exclusively FFI related).

**Comment by TimothyJones** [comment]
Sounds reasonable! Thanks for all your hard work!
